[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Blackpachamame/python/blob/main/Aula4/Aula_4_Estructuras_repetitivas.ipynb)

# Aula 4: Estructuras repetitivas

Vamos practicar el uso de estructuras de repetición como el "while" y el "for" a partir de algunas actividades. Ahora que estamos avanzando en el contenido, podemos hacer los desafíos más interesantes. ¡Para ello, trabajaremos en proyectos de código!

## Entrenando la programación

### Ejercicio 1

Escribe un programa que solicite dos números enteros e imprima todos los números enteros entre ellos.

In [8]:
inicio = int(input("Ingrese el primer número entero: "))
fin = int(input("Ingrese el segundo número entero: "))

if inicio < fin:
    for i in range(inicio + 1, fin):
        print(i)
elif fin < inicio:
    for i in range(fin + 1, inicio):
        print(i)
else:
    print("Los números son iguales")

Ingrese el primer número entero: 2
Ingrese el segundo número entero: 7
3
4
5
6


### Ejercicio 2

Escribe un programa para calcular cuántos días tomará que la colonia de una bacteria A supere o iguale a la colonia de una bacteria B, basado en tasas de crecimiento del 3% y 1.5%, respectivamente. Supón que la colonia A comienza con 4 elementos y B con 10.

In [10]:
dias = 0
bacterias_a = 4
bacterias_b = 10

while bacterias_a < bacterias_b:
    bacterias_a *= 1.03
    bacterias_b *= 1.015
    dias += 1

print("Tomará", dias, "días para que la colonia de A supere o iguale a la de B.")

Tomará 63 días para que la colonia de A supere o iguale a la de B.


### Ejercicio 3

Para procesar una cantidad de 15 datos de evaluaciones de usuarios de un servicio de la empresa, necesitamos verificar si las calificaciones son válidas. Por lo tanto, escribe un programa que recibirá calificaciones del 0 al 5 y verificará si son valores válidos. Si se ingresa una calificación superior a 5 o inferior a 0, se repetirá hasta que el usuario ingrese un valor válido.

In [3]:
for i in range(15):
    calificacion = float(input(f"Ingrese la calificación del usuario {i} (debe ser una nota del 0 al 5): "))
    while calificacion < 0 or calificacion > 5:
        calificacion = float(input("Calificación inválida. Intente nuevamente."))
print("Calificaciones cargadas correctamente.")

Ingrese la calificación del usuario 0 (debe ser una nota del 0 al 5): 4
Ingrese la calificación del usuario 1 (debe ser una nota del 0 al 5): 3
Ingrese la calificación del usuario 2 (debe ser una nota del 0 al 5): 5
Ingrese la calificación del usuario 3 (debe ser una nota del 0 al 5): 5
Ingrese la calificación del usuario 4 (debe ser una nota del 0 al 5): 4
Ingrese la calificación del usuario 5 (debe ser una nota del 0 al 5): 4
Ingrese la calificación del usuario 6 (debe ser una nota del 0 al 5): 6
Calificación inválida. Intente nuevamente.7
Calificación inválida. Intente nuevamente.5
Ingrese la calificación del usuario 7 (debe ser una nota del 0 al 5): 2
Ingrese la calificación del usuario 8 (debe ser una nota del 0 al 5): 3
Ingrese la calificación del usuario 9 (debe ser una nota del 0 al 5): 1
Ingrese la calificación del usuario 10 (debe ser una nota del 0 al 5): 3
Ingrese la calificación del usuario 11 (debe ser una nota del 0 al 5): 6
Calificación inválida. Intente nuevamente.5
In

### Ejercicio 4

Desarrolla un programa que lea un conjunto indefinido de temperaturas en grados Celsius y calcule su promedio. La lectura debe detenerse al ingresar el valor -273°C.

In [8]:
contador = 0
suma = 0
temperatura = float(input("Ingrese la temperatura en grados Celsius: "))

while temperatura != -273:
    suma += temperatura
    contador += 1
    temperatura = float(input("Ingrese la temperatura en grados Celsius: "))
if contador > 0:
    promedio = suma / contador
    print("El promedio de las temperaturas ingresadas es:", promedio)
else:
    print("No se ingresaron temperaturas válidas.")

Ingrese la temperatura en grados Celsius: 120
Ingrese la temperatura en grados Celsius: -23
Ingrese la temperatura en grados Celsius: -273
El promedio de las temperaturas ingresadas es: 48.5


### Ejercicio 5

Escribe un programa que calcule el factorial de un número entero proporcionado por el usuario. Recuerda que el factorial de un número entero es el producto de ese número por todos sus antecesores hasta llegar al número 1. Por ejemplo, el factorial de 5 es 5 x 4 x 3 x 2 x 1 = 120.

In [9]:
numero = int(input("Ingrese un número entero: "))
factorial = 1

if numero < 0:
    print("No se puede calcular el factorial de un número negativo.")
elif numero == 0 or numero == 1:
    print(f"El factorial de {numero} es {factorial}.")
else:
    for i in range(2, numero + 1):
        factorial *= i
    print(f"El factorial de {numero} es {factorial}.")

Ingrese un número entero: 5
El factorial de 5 es 120.


## Momento de los proyectos

### Ejercicio 6

Escribe un programa que genere la tabla de multiplicar de un número entero del 1 al 10, según la elección del usuario. Como ejemplo, para el número 2, la tabla de multiplicar debe mostrarse en el siguiente formato:

```
Tabla de multiplicar del 2:
2 x 1 = 2
2 x 2 = 4
[...]
2 x 10 = 20
```


### Ejercicio 7

Los números primos tienen diversas aplicaciones en Ciencia de Datos, como en criptografía y seguridad. Un número primo es aquel que es divisible solo por sí mismo y por 1. Por lo tanto, crea un programa que solicite un número entero y determine si es un número primo o no.

### Ejercicio 8

Vamos a comprender la distribución de edades de los pensionistas de una empresa de seguros. Escribe un programa que lea las edades de una cantidad no informada de clientes y muestre la distribución en los intervalos [0-25], [26-50], [51-75] y [76-100]. La entrada de datos se detendrá al ingresar un número negativo.

### Ejercicio 9

En una elección para la gerencia de una empresa con 20 empleados, hay cuatro candidatos. Escribe un programa que calcule al ganador de la elección. La votación se realizó de la siguiente manera:

Cada empleado votó por uno de los cuatro candidatos (representados por los números 1, 2, 3 y 4).

También se contaron los votos nulos (representados por el número 5) y los votos en blanco (representados por el número 6).

Al final de la votación, el programa debe mostrar el número total de votos para cada candidato, los votos nulos y los votos en blanco. Además, debe calcular y mostrar el porcentaje de votos nulos con respecto al total de votos y el porcentaje de votos en blanco con respecto al total de votos.